pkg/kubelet/volume/nfs/nfs.go

```diff
@@ -0,0 +1,208 @@
+	return &nfs{
+		volName:      spec.Name,
+		server:       spec.VolumeSource.NFS.Server,
+		exportPath:   spec.VolumeSource.NFS.SourcePath,
+		mountOptions: spec.VolumeSource.NFS.MountOptions,
+		mounter:      mounter,
+		podRef:       podRef,
+		plugin:       plugin,
+	}, nil
+}
+
+func (plugin *nfsPlugin) NewCleaner(volName string, podUID types.UID) (volume.Cleaner, error) {
+	return plugin.newCleanerInternal(volName, podUID, newNFSMounter())
+}
+
+func (plugin *nfsPlugin) newCleanerInternal(volName string, podUID types.UID, mounter nfsMountInterface) (volume.Cleaner, error) {
+	return &nfs{
+		volName:      volName,
+		server:       "",
+		exportPath:   "",
+		mountOptions: "",
+		mounter:      mounter,
+		podRef:       &api.ObjectReference{UID: podUID},
+		plugin:       plugin,
+	}, nil
+}
+
+// NFS volumes represent a bare host file or directory mount of an NFS export.
+type nfs struct {
+	volName      string
+	podRef       *api.ObjectReference
+	server       string
+	exportPath   string
+	mountOptions string
+	mounter      nfsMountInterface
+	plugin       *nfsPlugin
+}
+
+// SetUp attaches the disk and bind mounts to the volume path.
+func (nfs *nfs) SetUp() error {
+	return nfs.SetUpAt(nfs.GetPath())
+}
+
+func (nfs *nfs) SetUpAt(dir string) error {
+	mountpoint, err := mount.IsMountPoint(dir)
+	glog.V(4).Infof("NFS mount set up: %s %v %v", dir, mountpoint, err)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if mountpoint {
+		return nil
+	}
+	exportDir := nfs.exportPath
+	exists, err := exists(dir)
+	if err != nil {
+		glog.Errorf("Mount directory is invalid: %v, %v", dir, err)
+		return err
+	}
+	if !exists {
+		os.MkdirAll(dir, 0750)
+	}
+	err = nfs.mounter.Mount(nfs.server, exportDir, dir, nfs.mountOptions)
+	if err != nil {
+		mountpoint, mntErr := mount.IsMountPoint(dir)
+		if mntErr != nil {
+			glog.Errorf("IsMountpoint check failed: %v", mntErr)
+			return err
+		}
+		if mountpoint {
+			if mntErr = nfs.mounter.Unmount(dir); mntErr != nil {
+				glog.Errorf("Failed to unmount: %v", mntErr)
+				return err
+			}
+			mountpoint, mntErr := mount.IsMountPoint(dir)
+			if mntErr != nil {
+				glog.Errorf("IsMountpoint check failed: %v", mntErr)
+				return err
+			}
+			if mountpoint {
+				// This is very odd, we don't expect it.  We'll try again next sync loop.
+				glog.Errorf("%s is still mounted, despite call to unmount().  Will try again next sync loop.", nfs.GetPath())
+				return err
+			}
+		}
+		os.Remove(dir)
+		return err
+	}
+	return nil
+}
+
+func exists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	if err == nil {
+		return true, nil
+	}
+	if os.IsNotExist(err) {
+		return false, nil
+	}
+	return false, err
+}
+
+func (nfs *nfs) GetPath() string {
+	name := nfsPluginName
+	return nfs.plugin.host.GetPodVolumeDir(nfs.podRef.UID, volume.EscapePluginName(name), nfs.volName)
+}
+
+func (nfs *nfs) GetServer() string {
```

Comment:
Why do we have all of these public functions?  Do they make up some other interface?  It's not clear